text/0000-serialize-deserialize.md

```diff
@@ -0,0 +1,170 @@
+* Start Date: 2018-08-01
+* RFC PR: https://github.com/fusionjs/rfcs/pull/8
+
+# Summary
+
+This RFC describes a process for standardizing on an API for serializing and deserializing state between server and browser plugins.
+
+# Basic example
+
+```js
+import {SharedStateToken, createPlugin} from 'fusion-core';
+
+export default createPlugin({
+  deps: {
+    SharedState: SharedStateToken,
+  },
+  middleware: ({SharedState}) => {
+    return (ctx, next) => {
+      const {serialize, deserialize} = SharedState(ctx, 'some-id');
+      if (__BROWSER__) {
+        // load and parse serialized data from dom node
+        const data = deserialize();
+      } else {
+        // serialize json data into body
+        serialize({
+          some: 'json',
+        });
+      }
+      return next();
+    };
+  },
+});
+
+// in tests
+import {SharedStateToken, MockSharedState} from 'fusion-core';
+app.register(SharedStateToken, MockSharedState({
+  'some-id': {
+    some: 'json'
+  },
+  'other-data': {}
+}));
+```
+
+# Motivation
+
+It is a very common pattern in fusion to have some shared state between the server and
+the browser encapsulated into a plugin. The current model for this has users referencing
+`ctx.template` directly on the server, and querying the DOM directly from the browser.
+
+This can make testing difficult, especially when running browser tests since there is not
+an easy way to mock the deserialization other than each plugin exposing their own deserialization
+token. 
```

Comment:
Is this solution that bad (i.e. just having a token for each plugin)? I guess it depends on a few factors
- How frequently you want to mock the "universal" states for multiple separate plugins

Some advantages of per-plugin state mock/override tokens:
- Better type safety? I'm not totally sure but it seems like this might be difficult to type
- No need for string IDs (not having to remember string IDs, no typos)
- Plugins can't have *implicit* dependencies on other plugin's shared state (using string ids)

```js
import {FooStateToken} from "fusion-plugin-foo";
import {BarStateToken} from "fusion-plugin-bar";
import {BazStateToken} from "fusion-plugin-baz";

app.register(FooStateToken, /* something */);
app.register(BarStateToken, /* something */);
app.register(BazStateToken, /* something */);
```

I suppose this is more verbose, but could be remedied:

```js
import {FooStateToken} from "fusion-plugin-foo";
import {BarStateToken} from "fusion-plugin-bar";
import {BazStateToken} from "fusion-plugin-baz";

function mockState(app, state) {
  app.register(FooStateToken, state.foo);
  app.register(BarStateToken, state.bar);
  app.register(BazStateToken, state.baz);
}
```

Reply:
I don't think what you are proposing is correct or possible. Maybe you are missing some context from the RFC. The user doesn't provide the data, the plugin provides the data. So the data itself cannot come from the token. Maybe we should have a quick chat IRL to go over the doc